### PR TITLE
WIP: Dynesty checkpointing: With Hdf files as checkpoint files

### PR DESCRIPTION
**Pull request by sum33it**
To make the checkpoint filesystem uniform across samplers, I am starting this pull request. The sampler state is dumped to existing checkpoint pdf files using 'hickle' at regular intervals. Added two functions for this purpose in `pycbc.inference.io.dynesty`.

### CODE REVIEW

File: pycbc/inference/io/dynesty.py [hunk 1]
```diff
@@ -24,8 +24,20 @@
 """Provides IO for the dynesty sampler.
 """
 from .base_nested_sampler import BaseNestedSamplerFile
+import hickle
```
Comment (ahnitz) [inline pycbc/inference/io/dynesty.py:27]: Why are you using this and not the function Collin wrote?

Comment (sum33it) [reply pycbc/inference/io/dynesty.py]: Thanks for pointing. I fixed this issue. Also tested with a run with GW150914 and an analytical model. It seems to work now.

File: pycbc/inference/sampler/dynesty.py [hunk 2]
```diff
@@ -175,10 +231,23 @@ def finalize(self):
...
 
+    def getstate(self):
+        """Strips unserializable attributes of sampler
+        """
+        self_dict = self._sampler.__dict__.copy()
+        # Remove multiprocessing / mpi pool-related/ rstate attributes
+        # as they cannot be pickled
+        del self_dict['pool']
+        del self_dict['M']
+        del self_dict['rstate']
```
Comment (ahnitz) [inline pycbc/inference/sampler/dynesty.py:246]: This won't preserve the random seed for when you start back up from a resume. I think wee need to (1) set the random state explicitly. It might pickle on its own, which would be nice, but if it doens't, you'll have to explicitly preserve it and restore it on resume.

File: pycbc/inference/sampler/dynesty.py [hunk 3]
```diff
@@ -158,10 +200,24 @@ def from_config(cls, cp, model, output_file=None, nprocesses=1,
...
+        for fn in [self.checkpoint_file]:
+            with loadfile(fn, 'a') as fp:
+                fp.write_pickled_data_into_checkpoint_file(pickle_obj)
 
     def resume_from_checkpoint(self):
-        pass
+        for fn in [self.checkpoint_file]:
+            with loadfile(fn, 'r') as fp:
+                self._sampler = fp.read_pickled_data_from_checkpoint_file()
+                self._sampler.rstate = numpy.random
```
Comment (ahnitz) [inline pycbc/inference/sampler/dynesty.py:215]: See paired comment, this won't restore the random state to how it left off when checkpointed.

Comment (sum33it) [reply pycbc/inference/sampler/dynesty.py]: @ahnitz Now I have changed it and using functions written by @cdcapano: `write_random_state` and `read_random_state` to keep track of random state and restore it from previous checkpoint.

File: pycbc/inference/sampler/dynesty.py [hunk 4]
```diff
@@ -158,10 +200,35 @@ def from_config(cls, cp, model, output_file=None, nprocesses=1,
...
+            with loadfile(fn, 'a') as fp:
+                fp.write_pickled_data_into_checkpoint_file(pickle_obj)
+                fp.write_random_state()
 
     def resume_from_checkpoint(self):
-        pass
+        for fn in [self.checkpoint_file]:
+            with loadfile(fn, 'r') as fp:
+                self._sampler = fp.read_pickled_data_from_checkpoint_file()
+            self._sampler.rstate = numpy.random
```
Comment (ahnitz) [inline pycbc/inference/sampler/dynesty.py:216]: Why is this line here?

File: pycbc/inference/sampler/dynesty.py [hunk 5]
```diff
@@ -158,10 +200,35 @@ def from_config(cls, cp, model, output_file=None, nprocesses=1,
...
+            self._sampler.M = pool.map
+            self._sampler.pool = pool
+
+    def set_state_from_file(self, filename):
+        """Sets the state of the sampler back to the instance saved in a file.
+        """
+        with self.io(filename, 'r') as fp:
+            rstate = fp.read_random_state()
+        # set the numpy random state
+        numpy.random.set_state(rstate)
```
Comment (ahnitz) [inline pycbc/inference/sampler/dynesty.py:230]: Since dynesty has an explicitly rstate, why don't you just set that when you instatiate the sampler first? That way we don't have to rely on the state of the global numpy.

Comment (ahnitz) [reply pycbc/inference/sampler/dynesty.py:230]: Oh, is this a function that collin already had in place?

File: pycbc/inference/sampler/dynesty.py (hunk 4 above)
Comment (ahnitz) [reply pycbc/inference/sampler/dynesty.py:216]: Nvm, I think I see what you are doing.

File: pycbc/inference/io/__init__.py [hunk 6]
```diff
@@ -185,7 +192,7 @@ def check_integrity(filename):
                 _ = fp[group.format(param)][lastidx]
 
 
-def validate_checkpoint_files(checkpoint_file, backup_file):
+def validate_checkpoint_files(checkpoint_file, backup_file, sampler_name):
```
Comment (cdcapano) [inline pycbc/inference/io/__init__.py:195]: Don't make this dependent on the sampler name. That will be a major pain to keep up to date. Just have it be a keyword argument, like `check_nsamples=True`, then below replace the:
```
if sampler_name in ['dynesty', 'cpnest']:
    pass
else:
```
with:
```
if check_nsamples:
```
By making this a kwarg with default of True, you don't have to change any of the other codes, and just have to pass in `check_nsamples=False` into the call to validate_checkpoint_files for dynesty.

File: pycbc/inference/sampler/dynesty.py [hunk 7]
```diff
@@ -98,9 +112,35 @@ def __init__(self, model, nlive, nprocesses=1,
...
     def run(self):
-        self._sampler.run_nested(**self.run_kwds)
+        diff_niter = 1
+        if self.run_with_checkpoint is True:
+            if self.checkpoint_file:
+                try:
+                    self.resume_from_checkpoint()
+                    self.niter = self._sampler.results.niter
+                    logging.info('Successfully read from the checkpoint file')
+                except KeyError:
```
Comment (ahnitz) [inline pycbc/inference/sampler/dynesty.py:126]: Is a KeyError the only possible error?

File: pycbc/inference/sampler/dynesty.py [hunk 8]
```diff
@@ -80,12 +84,22 @@ def __init__(self, model, nlive, nprocesses=1,
...
+        self.maxcall = maxcall
+        self.checkpoint_time_interval = checkpoint_time_interval
         self.run_kwds = {} if run_kwds is None else run_kwds
         self.nlive = nlive
         self.names = model.sampling_params
         self.ndim = len(model.sampling_params)
         self.checkpoint_file = None
+        self.__getstate__ = self.getstate
+        self.niter = niter
+        if ('maxcall' or 'maxiter') in self.run_kwds:
```
Comment (ahnitz) [inline pycbc/inference/sampler/dynesty.py:98]: I'm confused by this. So there will be no checkpointing by default, if you don't set maxcall in run_kwds? It seems self.maxcall isn't actually used anyway. Why the inconsistency?

Comment (ahnitz) [reply pycbc/inference/sampler/dynesty.py:98]: Shouldn't checkpointing just be enabled if checkpoint_time_interval is set?

File: pycbc/inference/sampler/base.py [hunk 9]
```diff
@@ -175,7 +175,7 @@ def setup_output(sampler, output_file):
     # check if we have a good checkpoint and/or backup file
     logging.info("Looking for checkpoint file")
     checkpoint_valid = validate_checkpoint_files(checkpoint_file,
-                                                 backup_file)
+                                                 backup_file, sampler.name)
```
Comment (cdcapano) [inline pycbc/inference/sampler/base.py:178]: Continuing my suggestion above, add a `check_nsamples` keyword argument to `setup_output` (with default True), which you then pass to `validate_checkpoint_files`.

File: pycbc/inference/sampler/base_mcmc.py [hunk 10]
```diff
@@ -620,7 +620,7 @@ def checkpoint(self):
         # check validity
         logging.info("Validating checkpoint and backup files")
         checkpoint_valid = validate_checkpoint_files(
-            self.checkpoint_file, self.backup_file)
+            self.checkpoint_file, self.backup_file, self.name)
```
Comment (cdcapano) [inline pycbc/inference/sampler/base_mcmc.py:623]: With the keyword argument change, you can remove this change from base mcmc.

File: pycbc/inference/sampler/cpnest.py [hunk 11]
```diff
@@ -139,7 +139,7 @@ def finalize(self):
             self.write_results(fn)
         logging.info("Validating checkpoint and backup files")
         checkpoint_valid = validate_checkpoint_files(
-            self.checkpoint_file, self.backup_file)
+            self.checkpoint_file, self.backup_file, self.name)
```
Comment (cdcapano) [inline pycbc/inference/sampler/cpnest.py:142]: Here, pass in `check_nsamples=False`.

File: pycbc/inference/sampler/dynesty.py [hunk 12]
```diff
@@ -175,10 +242,23 @@ def finalize(self):
             self.write_results(fn)
         logging.info("Validating checkpoint and backup files")
         checkpoint_valid = validate_checkpoint_files(
-            self.checkpoint_file, self.backup_file)
+            self.checkpoint_file, self.backup_file, self.name)
```
Comment (cdcapano) [inline pycbc/inference/sampler/dynesty.py:245]: And here pass in `check_nsamples=False`

File: pycbc/inference/sampler/multinest.py [hunk 13]
```diff
@@ -327,7 +327,7 @@ def checkpoint(self):
                 f_p.write_niterations(self.niterations)
         logging.info("Validating checkpoint and backup files")
         checkpoint_valid = validate_checkpoint_files(
-            self.checkpoint_file, self.backup_file)
+            self.checkpoint_file, self.backup_file, self.name)
```
Comment (cdcapano) [inline pycbc/inference/sampler/multinest.py:330]: Here pass `check_nsamples=False`

File: pycbc/inference/sampler/dynesty.py (hunk 7 above)
Comment (sum33it) [reply pycbc/inference/sampler/dynesty.py:126]: At least in the testing it was the only error I encountered in various possible scenarios.

File: pycbc/inference/io/__init__.py (hunk 6 above)
Comment (sum33it) [reply pycbc/inference/io/__init__.py:195]: Ok. Thanks for the suggestion. Let me do this.

File: pycbc/inference/sampler/dynesty.py (hunk 7 above)
Comment (ahnitz) [reply pycbc/inference/sampler/dynesty.py:126]: If the key is there but is corrupt, this won't be the error. This is one of the examples, where a blanket except would be correct.

**Comment by ahnitz** [comment]
Closing in favor of #3354
